File: goldenflow/cli/watch.py
"""Watch a directory for file changes and auto-transform."""
from __future__ import annotations

import time
from pathlib import Path

from rich.console import Console

console = Console()


def watch_directory(
    path: Path,
    config_path: Path | None = None,
    output_dir: Path | None = None,
    interval: float = 2.0,
) -> None:
    """Watch a directory and transform new/changed CSV files."""
    from goldenflow.config.loader import load_config
    from goldenflow.config.schema import GoldenFlowConfig
    from goldenflow.engine.transformer import TransformEngine

    cfg = load_config(config_path) if config_path else GoldenFlowConfig()
    engine = TransformEngine(config=cfg)
    out = output_dir or path

    seen: dict[str, float] = {}  # path -> last modified time

    console.print(f"[bold]Watching[/bold] {path} (interval: {interval}s)")
    console.print("[dim]Press Ctrl+C to stop[/dim]\n")

    try:
        while True:
            for f in path.glob("*.csv"):
                mtime = f.stat().st_mtime
                if str(f) not in seen or seen[str(f)] < mtime:
                    # Skip already-transformed files
                    if "_transformed" in f.name:
                        continue
                    console.print(f"[cyan]Detected:[/cyan] {f.name}")
                    try:
                        result = engine.transform_file(f, output_dir=out)
                        console.print(
                            f"  [green]Transformed:[/green] {len(result.manifest.records)} transforms applied"
                        )
                    except Exception as e:
                        console.print(f"  [red]Error:[/red] {e}")
                    seen[str(f)] = mtime
            time.sleep(interval)
    except KeyboardInterrupt:
        console.print("\n[dim]Watch stopped.[/dim]")

File: packages/goldenflow-js/src/node/watch.ts
import { readdirSync, statSync } from "node:fs";
import { join, extname } from "node:path";
import { readFile, writeFile } from "./connectors/file.js";
import { TransformEngine } from "../core/engine/transformer.js";
import { loadConfigFromString } from "../core/config/loader.js";
import { makeConfig } from "../core/types.js";
import { readFileSync } from "node:fs";

export function watchDirectory(
  dirPath: string,
  options: { configPath?: string; outputDir?: string; interval?: number } = {},
): void {
  const interval = options.interval ?? 2.0;
  const cfg = options.configPath
    ? loadConfigFromString(readFileSync(options.configPath, "utf-8"))
    : makeConfig();
  const engine = new TransformEngine(cfg);
  const outDir = options.outputDir ?? dirPath;
  const seen = new Map<string, number>(); // path -> mtime

  console.log(`Watching ${dirPath} (interval: ${interval}s)`);
  console.log("Press Ctrl+C to stop\n");

  const poll = () => {
    try {
      const files = readdirSync(dirPath).filter(f => extname(f).toLowerCase() === ".csv");
      for (const file of files) {
        if (file.includes("_transformed")) continue;
        const fullPath = join(dirPath, file);
        const mtime = statSync(fullPath).mtimeMs;
        if (!seen.has(fullPath) || seen.get(fullPath)! < mtime) {
          console.log(`Detected: ${file}`);
          try {
            const rows = readFile(fullPath);
            const result = engine.transformDf(rows, fullPath);
            const stem = file.replace(extname(file), "");
            writeFile(result.rows as Record<string, unknown>[], join(outDir, `${stem}_transformed.csv`));
            console.log(`  Transformed: ${result.manifest.records.length} transforms applied`);
          } catch (e) {
            console.error(`  Error: ${e instanceof Error ? e.message : String(e)}`);
          }
          seen.set(fullPath, mtime);
        }
      }
    } catch (e) {
      console.error(`Watch error: ${e instanceof Error ? e.message : String(e)}`);
    }
  };

  poll();
  const timer = setInterval(poll, interval * 1000);
  process.on("SIGINT", () => { clearInterval(timer); console.log("\nWatch stopped."); process.exit(0); });
}
